src/include/concepts.h

```diff
@@ -36,78 +36,196 @@
-      });
-  { t.data() } -> std::convertible_to<typename T::value_type*>;
-  { num_features(t) } -> std::convertible_to<std::size_t>;
+concept dimensionable = requires(const T& t) {
+  { dimension(t) } -> semi_integral;
 };
 
-template <typename T>
-concept query_vector = feature_vector<T>;
-
-template <typename T>
-concept vector_database = requires(T t) {
-  typename T::value_type;
-  //  typename T::index_type;
-  //  typename T::size_type;
-  //  typename T::reference;
-  { t.size() } -> std::convertible_to<std::size_t>;
-  { t.rank() } -> std::convertible_to<std::size_t>;
-  { t[0] } -> std::convertible_to<std::span<typename T::value_type>>;
-  { t(0, 0) } -> std::convertible_to<typename T::value_type>;
-  { t.data() } -> std::convertible_to<typename T::value_type*>;
-  { t.rank() == 2 };
-  { raveled(t) } -> std::convertible_to<std::span<typename T::value_type>>;
+template <class T>
+concept vectorable = requires(const T& t) {
+  { num_vectors(t) } -> semi_integral;
 };
 
-template <typename T>
-concept query_set = vector_database<T>;
+template <class T>
+concept partitionable = requires(const T& t) {
+  //  { num_partitions(t) } -> semi_integral;
+  { t.indices() };
+  { t.ids() };
+};
+
+// ----------------------------------------------------------------------------
+// feature_vector concept
+// ----------------------------------------------------------------------------
+
+template <typename R>
+concept feature_vector =
+    std::ranges::random_access_range<R> && /* std::ranges::sized_range<R> && */
+    std::ranges::contiguous_range<R> && dimensionable<R> &&
+    (subscriptable_range<R> || callable_range<R>);
+
+template <class R>
+concept query_vector = feature_vector<R>;
+
+// ----------------------------------------------------------------------------
+// feature_vector_array concept
+//
+// A vector range is a range of feature vectors.
+// We will want at least these kinds:
+//   contiguous -- all the vectors are stored in a single contiguous array
+//   partitioned -- the range is a range of partitions, each of which is a
+//                  contiguous range of vectors
+//   distributed?-- the range is a range of partitions, each of which is a
+//
+// @todo operator()(size_t, size_t) ?
+// ----------------------------------------------------------------------------
+template <class D>
+concept feature_vector_array =
+    // feature_vector<inner_range_t<D>> &&
+    // std::ranges::random_access_range<D> && /* std::ranges::sized_range<D> &&
+    // */ subscriptable_range<D> && requires(D d, const
+    // std::iter_difference_t<std::ranges::iterator_t<D>> n) {
+    requires(D d, size_t n) {
+      //      typename D::feature_type;
+      { num_vectors(d) } -> semi_integral;
+      { dimension(d) } -> semi_integral;
+      //{
+      //        d[n]
+      //      } ->
+      //      std::same_as<std::iter_reference_t<std::ranges::iterator_t<D>>>;
+      { d[n] } -> feature_vector;  // Maybe redundant
```

Comment:
Did you determine whether it's redundant or not? Would be good to either remove this or add more context to the comment, as a future reader (i.e. me 😅) will be quite confused about this.

Reply:
I have not determined that.  Determining this will take some experimentations, which we should defer since it is not harmful at any rate.